### PR TITLE
Move to minimum of PHP 7.1

### LINKED ISSUE #194: Add type declarations throughout, including ` : void` (php 7.1)
Bump version to PHP 7.1 and add type declarations for parameters and return values throughout the project, including ` : void`.

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #194

### CODE REVIEW

**Comment by asgrim** [comment]
No idea what's going on with Scrutinizer, but otherwise we're good here for review I think!

**Comment by asgrim** [comment]
I retried the Scrutinizer build, all green now!

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ PHP_BrowscapINI
 ergebnis.txt
 tests/data/*
 !tests/data/.gitkeep
+.php_cs.cache
```
Comment (jaydiablo) [inline .gitignore]: You added this here, but I don't see the `setUsingCache` config set in the `.php_cs` file. Was php-cs-fixer creating the cache file even with that disabled (maybe it's default now)?

Thoughts on enabling and using the cache?  Generally I like having the cache as it speeds up the php-cs-fixer run considerably.

File: .php_cs [hunk 2]
```diff
@@ -8,84 +10,64 @@ $finder = Symfony\CS\Finder\DefaultFinder::create()
...
+        'strict_comparison' => true,
+        'ternary_operator_spaces' => true,
+        'ternary_to_null_coalescing' => true,
+        'trailing_comma_in_multiline_array' => true,
+        'trim_array_spaces' => true,
+        'unary_operator_spaces' => true,
+        'visibility_required' => true,
+        'whitespace_after_comma_in_array' => true,
+    ])
+    ->setFinder($finder)
```
Comment (jaydiablo) [inline .php_cs]: ` ->setUsingCache(true)` to enable cache if desired (and not default behavior).

File: src/Formatter/LegacyFormatter.php [hunk 3]
```diff
@@ -79,22 +45,18 @@ public function __construct($options = [])
...
-        return $this;
     }
 
     /**
      * Gets the data (in the preferred format)
      *
      * @return \stdClass
      */
-    public function getData()
+    public function getData() : ?\stdClass
```
Comment (jaydiablo) [inline src/Formatter/LegacyFormatter.php]: Should this one be nullable? Seems that it always returns the `$output` var (which is a stdClass) no?

File: src/Formatter/PhpGetBrowser.php [hunk 4]
```diff
@@ -111,34 +77,31 @@ class PhpGetBrowser implements FormatterInterface
...
-        return $this;
     }
 
     /**
      * Gets the data (in the preferred format)
      *
      * @return \stdClass
      */
-    public function getData()
+    public function getData() : ?\stdClass
```
Comment (jaydiablo) [inline src/Formatter/PhpGetBrowser.php]: Shouldn't be nullable?

File: src/Parser/Helper/Pattern.php [hunk 5]
```diff
@@ -1,47 +1,17 @@
...
- * @author     Thomas Müller <t_mueller_stolzenhain@yahoo.de>
- * @copyright  Copyright (c) 1998-2015 Browser Capabilities Project
- * @version    3.0
- * @license    http://www.opensource.org/licenses/MIT MIT License
- * @link       https://github.com/browscap/browscap-php/
  */
-class Pattern
+final class Pattern
 {
+    private function __construct()
```
Comment (jaydiablo) [inline src/Parser/Helper/Pattern.php]: Is the empty constructor necessary? Just being explicit?

File: src/Formatter/LegacyFormatter.php (hunk 3 above)
Comment (asgrim) [inline src/Formatter/LegacyFormatter.php]: I agree, I don't think it should be `null`able, but there is currently test(s) that pass back `null` from mocks of these. Created #201 to look into this further, but I'll leave as is for this PR.

File: src/Formatter/PhpGetBrowser.php (hunk 4 above)
Comment (asgrim) [inline src/Formatter/PhpGetBrowser.php]: See #201

File: src/Parser/Helper/Pattern.php (hunk 5 above)
Comment (asgrim) [inline src/Parser/Helper/Pattern.php]: It's `private` which indicates the `Pattern` class cannot be instantiated (it is intended to be used statically)

File: .gitignore (hunk 1 above)
Comment (asgrim) [inline .gitignore]: It was created on my machine. The cache can exist on your local machine if gitignored, but I'd prefer not to have it committed to the repo tbh.

Comment (jaydiablo) [inline .gitignore]: Yeah, pros and cons to that. We commit it in our projects which can cause merge conflicts that can really only be resolved by blowing away the file and re-running php-cs-fixer.  But I believe when the PHP version is different the cache is invalidated, so unless all developers are using same PHP version, doesn't seem to make much sense to commit it.

File: .php_cs (hunk 2 above)
Comment (jaydiablo) [inline .php_cs]: Ah, good to know.

**Comment by asgrim** [comment]
Thanks @mimmi20 and @jaydiablo !

**Comment by jaydiablo** [comment]
This will require that the useragent tests in browscap/browscap use PHP 7.1 now yeah?  I believe it has `dev-master` as the composer dependency for browscap/browscap-php.

**Comment by asgrim** [comment]
Ah yes, we should probably lock that down to 3.1.0 for now. Reported browscap/browscap#1335

**Pull request merged**